internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +100,210 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+			conditions = append(conditions, invalidPolicyCondition)
+		}
+		break
+	case v1alpha1.GitHubAllowedAuthOrganizationsPolicyOnlyUsersFromAllowedOrganizations:
+		if len(upstream.Spec.AllowAuthentication.Organizations.Allowed) == 0 {
+			conditions = append(conditions, invalidPolicyCondition)
+		}
+		break
+	default:
+		conditions = append(conditions, invalidPolicyCondition)
+	}
+
+	result := &upstreamgithub.ProviderConfig{
+		Name:               upstream.Name,
+		ResourceUID:        upstream.UID,
+		Host:               fmt.Sprintf("https://" + host),
+		GroupNameAttribute: *upstream.Spec.Claims.Groups,
+		UsernameAttribute:  *upstream.Spec.Claims.Username,
+		Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+		},
+		AllowedOrganizations:    upstream.Spec.AllowAuthentication.Organizations.Allowed,
+		OrganizationLoginPolicy: policy,
+		AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", host),
 	}
 
 	// TODO: once we firm up the proposal doc & merge, then firm up the CRD & merge, we can fill out these validations.
 	// The critical pattern to maintain is that every run of the sync loop will populate the exact number of the exact
 	// same set of conditions.  Conditions depending on other conditions should get Status:  metav1.ConditionUnknown, or
 	// Status:  metav1.ConditionFalse, never be omitted.
-	conditions := []*metav1.Condition{
-		// we may opt to split this up into smaller validation functions.
-		// Each function should be responsible for validating one logical unit and setting one condition.
-		// c.validateGitHubAPI(),
-		// c.validateClaims(),
-		// c.validateAllowedAuthentication(),
-		// c.validateClient(),
+	if !hasConditionByType(conditions, TLSConfigurationValid) {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: fmt.Sprintf("tls configuration is valid for spec.githubAPI.host (%q)", host),
+		})
+	}
+	if !hasConditionByType(conditions, OrganizationsPolicyValid) {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    OrganizationsPolicyValid,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: "allowAuthentication policy for organizations is valid",
+		})
+	}
+	if !hasConditionByType(conditions, GitHubConnectionValid) {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: fmt.Sprintf("spec.githubAPI.host (%q) is reachable and TLS verification succeeds", host),
+		})
+	}
+
+	hasError := conditionsutil.HadErrorCondition(conditions)
+	applicationErr = c.updateStatus(ctx.Context, upstream, conditions)
+	if hasError {
+		return nil, applicationErr
+	}
+	return result, applicationErr
+}
+
+func hasConditionByType(conditions []*metav1.Condition, t string) bool {
```

Comment:
Bet this can move into [Condition Utils](https://github.com/vmware-tanzu/pinniped/blob/main/internal/controller/conditionsutil/conditions_util.go)

Reply:
it was actually removed